v2/routing_test.go

```diff
@@ -419,23 +474,493 @@ func TestDHT_FindProvidersAsync_datastore_error(t *testing.T) {
+	require.NoError(t, err)
+
+	_, validValue := makeIPNSKeyValue(t, clk, priv, 1, time.Hour)
+	_, worseValue := makeIPNSKeyValue(t, clk, priv, 0, time.Hour)
+	key, betterValue := makeIPNSKeyValue(t, clk, priv, 2, time.Hour) // higher sequence number means better value
+
+	top := NewTopology(t)
+	d1 := top.AddServer(cfg)
+	d2 := top.AddServer(cfg)
+	d3 := top.AddServer(cfg)
+	d4 := top.AddServer(cfg)
+	d5 := top.AddServer(cfg)
+
+	fmt.Println(d1.host.ID().ShortString(), "Host 1")
+	fmt.Println(d2.host.ID().ShortString(), "Host 2")
+	fmt.Println(d3.host.ID().ShortString(), "Host 3")
+	fmt.Println(d4.host.ID().ShortString(), "Host 4")
+	fmt.Println(d5.host.ID().ShortString(), "Host 5")
+
+	top.ConnectChain(ctx, d1, d2, d3, d4, d5)
+
+	err = d3.putValueLocal(ctx, key, validValue)
+	require.NoError(t, err)
+
+	err = d4.putValueLocal(ctx, key, worseValue)
+	require.NoError(t, err)
+
+	err = d5.putValueLocal(ctx, key, betterValue)
+	require.NoError(t, err)
+
+	val, err := d1.GetValue(ctx, key)
+	assert.NoError(t, err)
+	assert.Equal(t, betterValue, val)
+}
+
+func TestDHT_GetValue_returns_context_error(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+	d := newTestDHT(t)
+
+	cancelledCtx, cancel := context.WithCancel(ctx)
+	cancel()
+
+	_, err := d.GetValue(cancelledCtx, "/"+namespaceIPNS+"/some-key")
+	assert.ErrorIs(t, err, context.Canceled)
+}
+
+func TestDHT_GetValue_returns_not_found_error(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+	d := newTestDHT(t)
+
+	valueChan, err := d.GetValue(ctx, "/"+namespaceIPNS+"/some-key")
+	assert.ErrorIs(t, err, routing.ErrNotFound)
+	assert.Nil(t, valueChan)
+}
+
+// assertClosed triggers a test failure if the given channel was not closed but
+// carried more values or a timeout occurs (given by the context).
+func assertClosed[T any](t testing.TB, ctx context.Context, c <-chan T) {
+	t.Helper()
+
 	select {
-	case _, more := <-out:
+	case _, more := <-c:
 		assert.False(t, more)
 	case <-ctx.Done():
-		t.Fatal("timeout")
+		t.Fatal("timeout closing channel")
 	}
 }
+
+func readItem[T any](t testing.TB, ctx context.Context, c <-chan T) T {
+	t.Helper()
+
+	select {
+	case val, more := <-c:
+		require.True(t, more, "channel closed unexpectedly")
+		return val
+	case <-ctx.Done():
+		t.Fatal("timeout reading item")
+		return *new(T)
+	}
+}
+
+func TestDHT_SearchValue_simple(t *testing.T) {
+	// Test setup:
+	// There is just one other server that returns a valid value.
+	ctx := kadtest.CtxShort(t)
+
+	key, v := makePkKeyValue(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+
+	top.Connect(ctx, d1, d2)
+
+	err := d2.putValueLocal(ctx, key, v)
+	require.NoError(t, err)
+
+	valChan, err := d1.SearchValue(ctx, key)
+	require.NoError(t, err)
+
+	val := readItem(t, ctx, valChan)
+	assert.Equal(t, v, val)
+
+	assertClosed(t, ctx, valChan)
+}
+
+func TestDHT_SearchValue_returns_best_values(t *testing.T) {
+	// Test setup:
+	// d2 returns no value
```

Comment:
Can you mention that the peers are connected in a chain so that the query will progress from d2, d3, d4, d5. It took a while for me to realise that the worse value is always rejected because it is always received after the valid value